{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Phosphorylation Site Data Cleaning\n",
    "\n",
    "Having obtained the raw data for kinase phosphorylation sites from PhosphoSitePlus as two Excel spreadsheets: Kinase_Substrate_Dataset and Phosphorylation_site_dataset, we now have to filter, clean and combine both datasets to get it into a form suitable to be used as a table for our database. First we'll work on filtering the Kinase_Substrate_Dataset as majority of the data for our table is going to be in this dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Open both of our raw data files in pandas and converting them into panda dataframes\n",
    "df = pd.read_csv('Kinase_Substrate_Dataset.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>KINASE</th>\n",
       "      <th>KIN_ACC_ID</th>\n",
       "      <th>KIN_ORGANISM</th>\n",
       "      <th>SUBSTRATE</th>\n",
       "      <th>SUB_GENE_ID</th>\n",
       "      <th>SUB_ACC_ID</th>\n",
       "      <th>SUB_GENE</th>\n",
       "      <th>SUB_ORGANISM</th>\n",
       "      <th>SUB_MOD_RSD</th>\n",
       "      <th>SITE_GRP_ID</th>\n",
       "      <th>SITE_+/-7_AA</th>\n",
       "      <th>DOMAIN</th>\n",
       "      <th>IN_VIVO_RXN</th>\n",
       "      <th>IN_VITRO_RXN</th>\n",
       "      <th>CST_CAT#</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Pak2</td>\n",
       "      <td>PAK2</td>\n",
       "      <td>Q64303</td>\n",
       "      <td>rat</td>\n",
       "      <td>MEK1</td>\n",
       "      <td>170851.0</td>\n",
       "      <td>Q01986</td>\n",
       "      <td>Map2k1</td>\n",
       "      <td>rat</td>\n",
       "      <td>S298</td>\n",
       "      <td>448284</td>\n",
       "      <td>RtPGRPLsSYGMDSR</td>\n",
       "      <td>Pkinase</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td>9128; 98195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Pak2</td>\n",
       "      <td>PAK2</td>\n",
       "      <td>Q64303</td>\n",
       "      <td>rat</td>\n",
       "      <td>PRKD1</td>\n",
       "      <td>85421.0</td>\n",
       "      <td>Q9WTQ1</td>\n",
       "      <td>Prkd1</td>\n",
       "      <td>rat</td>\n",
       "      <td>S203</td>\n",
       "      <td>449896</td>\n",
       "      <td>GVRRRRLsNVsLTGL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Pak2</td>\n",
       "      <td>PAK2</td>\n",
       "      <td>Q64303</td>\n",
       "      <td>rat</td>\n",
       "      <td>prolactin</td>\n",
       "      <td>5617.0</td>\n",
       "      <td>P01236</td>\n",
       "      <td>PRL</td>\n",
       "      <td>human</td>\n",
       "      <td>S207</td>\n",
       "      <td>451732</td>\n",
       "      <td>LHCLRRDsHKIDNYL</td>\n",
       "      <td>Hormone_1</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Pak2</td>\n",
       "      <td>PAK2</td>\n",
       "      <td>Q64303</td>\n",
       "      <td>rat</td>\n",
       "      <td>prolactin</td>\n",
       "      <td>24683.0</td>\n",
       "      <td>P01237</td>\n",
       "      <td>Prl</td>\n",
       "      <td>rat</td>\n",
       "      <td>S206</td>\n",
       "      <td>451732</td>\n",
       "      <td>IRCLRRDsHKVDNYL</td>\n",
       "      <td>Hormone_1</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>EIF2AK1</td>\n",
       "      <td>HRI</td>\n",
       "      <td>Q9BQI3</td>\n",
       "      <td>human</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>54318.0</td>\n",
       "      <td>P68101</td>\n",
       "      <td>Eif2s1</td>\n",
       "      <td>rat</td>\n",
       "      <td>S52</td>\n",
       "      <td>447635</td>\n",
       "      <td>MILLSELsRRRIRSI</td>\n",
       "      <td>S1</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td>3597; 9721; 3398; 5199</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      GENE KINASE KIN_ACC_ID KIN_ORGANISM   SUBSTRATE  SUB_GENE_ID SUB_ACC_ID  \\\n",
       "0     Pak2   PAK2     Q64303          rat        MEK1     170851.0     Q01986   \n",
       "1     Pak2   PAK2     Q64303          rat       PRKD1      85421.0     Q9WTQ1   \n",
       "2     Pak2   PAK2     Q64303          rat   prolactin       5617.0     P01236   \n",
       "3     Pak2   PAK2     Q64303          rat   prolactin      24683.0     P01237   \n",
       "4  EIF2AK1    HRI     Q9BQI3        human  eIF2-alpha      54318.0     P68101   \n",
       "\n",
       "  SUB_GENE SUB_ORGANISM SUB_MOD_RSD  SITE_GRP_ID     SITE_+/-7_AA     DOMAIN  \\\n",
       "0   Map2k1          rat        S298       448284  RtPGRPLsSYGMDSR    Pkinase   \n",
       "1    Prkd1          rat        S203       449896  GVRRRRLsNVsLTGL        NaN   \n",
       "2      PRL        human        S207       451732  LHCLRRDsHKIDNYL  Hormone_1   \n",
       "3      Prl          rat        S206       451732  IRCLRRDsHKVDNYL  Hormone_1   \n",
       "4   Eif2s1          rat         S52       447635  MILLSELsRRRIRSI         S1   \n",
       "\n",
       "  IN_VIVO_RXN IN_VITRO_RXN                CST_CAT#  \n",
       "0                        X             9128; 98195  \n",
       "1           X                                  NaN  \n",
       "2                        X                     NaN  \n",
       "3                        X                     NaN  \n",
       "4                        X  3597; 9721; 3398; 5199  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#To see what our dataframe looks like and to ensure pandas has loaded in our data correctly.\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, major thing we notice here is that under the Kinase Organism and Substrate Organism columns there are other species of animals included. As we are only interested in human kinases we filter accordingly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>KINASE</th>\n",
       "      <th>KIN_ACC_ID</th>\n",
       "      <th>KIN_ORGANISM</th>\n",
       "      <th>SUBSTRATE</th>\n",
       "      <th>SUB_GENE_ID</th>\n",
       "      <th>SUB_ACC_ID</th>\n",
       "      <th>SUB_GENE</th>\n",
       "      <th>SUB_ORGANISM</th>\n",
       "      <th>SUB_MOD_RSD</th>\n",
       "      <th>SITE_GRP_ID</th>\n",
       "      <th>SITE_+/-7_AA</th>\n",
       "      <th>DOMAIN</th>\n",
       "      <th>IN_VIVO_RXN</th>\n",
       "      <th>IN_VITRO_RXN</th>\n",
       "      <th>CST_CAT#</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>EIF2AK1</td>\n",
       "      <td>HRI</td>\n",
       "      <td>Q9BQI3</td>\n",
       "      <td>human</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>1965.0</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>human</td>\n",
       "      <td>S52</td>\n",
       "      <td>447635</td>\n",
       "      <td>MILLsELsRRRIRsI</td>\n",
       "      <td>S1</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td>3597; 9721; 3398; 5199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>EIF2AK1</td>\n",
       "      <td>HRI</td>\n",
       "      <td>Q9BQI3</td>\n",
       "      <td>human</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>1965.0</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>human</td>\n",
       "      <td>S49</td>\n",
       "      <td>450210</td>\n",
       "      <td>IEGMILLsELsRRRI</td>\n",
       "      <td>S1</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>PRKCD</td>\n",
       "      <td>PKCD</td>\n",
       "      <td>Q05655</td>\n",
       "      <td>human</td>\n",
       "      <td>HDAC5</td>\n",
       "      <td>10014.0</td>\n",
       "      <td>Q9UQL6</td>\n",
       "      <td>HDAC5</td>\n",
       "      <td>human</td>\n",
       "      <td>S259</td>\n",
       "      <td>447995</td>\n",
       "      <td>FPLRkTAsEPNLKVR</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>X</td>\n",
       "      <td>3443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>PRKCD</td>\n",
       "      <td>PKCD</td>\n",
       "      <td>Q05655</td>\n",
       "      <td>human</td>\n",
       "      <td>PTPRA iso2</td>\n",
       "      <td>5786.0</td>\n",
       "      <td>P18433-2</td>\n",
       "      <td>PTPRA</td>\n",
       "      <td>human</td>\n",
       "      <td>S204</td>\n",
       "      <td>447612</td>\n",
       "      <td>PLLARSPsTNRKYPP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>PRKCD</td>\n",
       "      <td>PKCD</td>\n",
       "      <td>Q05655</td>\n",
       "      <td>human</td>\n",
       "      <td>hnRNP K</td>\n",
       "      <td>3190.0</td>\n",
       "      <td>P61978</td>\n",
       "      <td>HNRNPK</td>\n",
       "      <td>human</td>\n",
       "      <td>S302</td>\n",
       "      <td>457408</td>\n",
       "      <td>GrGGrGGsrArNLPL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X</td>\n",
       "      <td>X</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       GENE KINASE KIN_ACC_ID KIN_ORGANISM   SUBSTRATE  SUB_GENE_ID  \\\n",
       "6   EIF2AK1    HRI     Q9BQI3        human  eIF2-alpha       1965.0   \n",
       "7   EIF2AK1    HRI     Q9BQI3        human  eIF2-alpha       1965.0   \n",
       "10    PRKCD   PKCD     Q05655        human       HDAC5      10014.0   \n",
       "11    PRKCD   PKCD     Q05655        human  PTPRA iso2       5786.0   \n",
       "12    PRKCD   PKCD     Q05655        human     hnRNP K       3190.0   \n",
       "\n",
       "   SUB_ACC_ID SUB_GENE SUB_ORGANISM SUB_MOD_RSD  SITE_GRP_ID     SITE_+/-7_AA  \\\n",
       "6      P05198   EIF2S1        human         S52       447635  MILLsELsRRRIRsI   \n",
       "7      P05198   EIF2S1        human         S49       450210  IEGMILLsELsRRRI   \n",
       "10     Q9UQL6    HDAC5        human        S259       447995  FPLRkTAsEPNLKVR   \n",
       "11   P18433-2    PTPRA        human        S204       447612  PLLARSPsTNRKYPP   \n",
       "12     P61978   HNRNPK        human        S302       457408  GrGGrGGsrArNLPL   \n",
       "\n",
       "   DOMAIN IN_VIVO_RXN IN_VITRO_RXN                CST_CAT#  \n",
       "6      S1                        X  3597; 9721; 3398; 5199  \n",
       "7      S1                        X                     NaN  \n",
       "10    NaN                        X                    3443  \n",
       "11    NaN           X                                  NaN  \n",
       "12    NaN           X            X                     NaN  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Filtering the dataframe for values where both Kinase and Substrate organisms are both human and checking to make sure this has\n",
    "#been done\n",
    "df = df[(df.KIN_ORGANISM=='human') & (df.SUB_ORGANISM=='human')]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, there are a number of columns in the dataframe that are not useful to us and so can be removed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>KINASE</th>\n",
       "      <th>KIN_ACC_ID</th>\n",
       "      <th>SUBSTRATE</th>\n",
       "      <th>SUB_ACC_ID</th>\n",
       "      <th>SUB_GENE</th>\n",
       "      <th>SUB_MOD_RSD</th>\n",
       "      <th>SITE_+/-7_AA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>HRI</td>\n",
       "      <td>Q9BQI3</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>S52</td>\n",
       "      <td>MILLsELsRRRIRsI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>HRI</td>\n",
       "      <td>Q9BQI3</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>S49</td>\n",
       "      <td>IEGMILLsELsRRRI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>PKCD</td>\n",
       "      <td>Q05655</td>\n",
       "      <td>HDAC5</td>\n",
       "      <td>Q9UQL6</td>\n",
       "      <td>HDAC5</td>\n",
       "      <td>S259</td>\n",
       "      <td>FPLRkTAsEPNLKVR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>PKCD</td>\n",
       "      <td>Q05655</td>\n",
       "      <td>PTPRA iso2</td>\n",
       "      <td>P18433-2</td>\n",
       "      <td>PTPRA</td>\n",
       "      <td>S204</td>\n",
       "      <td>PLLARSPsTNRKYPP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>PKCD</td>\n",
       "      <td>Q05655</td>\n",
       "      <td>hnRNP K</td>\n",
       "      <td>P61978</td>\n",
       "      <td>HNRNPK</td>\n",
       "      <td>S302</td>\n",
       "      <td>GrGGrGGsrArNLPL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   KINASE KIN_ACC_ID   SUBSTRATE SUB_ACC_ID SUB_GENE SUB_MOD_RSD  \\\n",
       "6     HRI     Q9BQI3  eIF2-alpha     P05198   EIF2S1         S52   \n",
       "7     HRI     Q9BQI3  eIF2-alpha     P05198   EIF2S1         S49   \n",
       "10   PKCD     Q05655       HDAC5     Q9UQL6    HDAC5        S259   \n",
       "11   PKCD     Q05655  PTPRA iso2   P18433-2    PTPRA        S204   \n",
       "12   PKCD     Q05655     hnRNP K     P61978   HNRNPK        S302   \n",
       "\n",
       "       SITE_+/-7_AA  \n",
       "6   MILLsELsRRRIRsI  \n",
       "7   IEGMILLsELsRRRI  \n",
       "10  FPLRkTAsEPNLKVR  \n",
       "11  PLLARSPsTNRKYPP  \n",
       "12  GrGGrGGsrArNLPL  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Remove columns in the df that aren't useful to us by their heading name and checking to make sure this has been done.\n",
    "df = df.drop(['GENE','KIN_ORGANISM', 'SUB_GENE_ID', 'SUB_ORGANISM', 'SITE_GRP_ID', 'DOMAIN', 'IN_VIVO_RXN', 'IN_VITRO_RXN', \n",
    "              'CST_CAT#'],axis=1)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As SQLite and SQLAlchemy has some specific syntax when referring to specific columns we rename our columns so that we have no errors with the database and web app integration, in addition to clarity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>KINASE_GENE_NAME</th>\n",
       "      <th>KIN_UNIPROT_ID</th>\n",
       "      <th>SUBSTRATE_NAME</th>\n",
       "      <th>SUB_UNIPROT_ID</th>\n",
       "      <th>SUB_GENE_NAME</th>\n",
       "      <th>SUB_MOD_RSD</th>\n",
       "      <th>SITE_7_AA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>HRI</td>\n",
       "      <td>Q9BQI3</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>S52</td>\n",
       "      <td>MILLsELsRRRIRsI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>HRI</td>\n",
       "      <td>Q9BQI3</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>S49</td>\n",
       "      <td>IEGMILLsELsRRRI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>PKCD</td>\n",
       "      <td>Q05655</td>\n",
       "      <td>HDAC5</td>\n",
       "      <td>Q9UQL6</td>\n",
       "      <td>HDAC5</td>\n",
       "      <td>S259</td>\n",
       "      <td>FPLRkTAsEPNLKVR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>PKCD</td>\n",
       "      <td>Q05655</td>\n",
       "      <td>PTPRA iso2</td>\n",
       "      <td>P18433-2</td>\n",
       "      <td>PTPRA</td>\n",
       "      <td>S204</td>\n",
       "      <td>PLLARSPsTNRKYPP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>PKCD</td>\n",
       "      <td>Q05655</td>\n",
       "      <td>hnRNP K</td>\n",
       "      <td>P61978</td>\n",
       "      <td>HNRNPK</td>\n",
       "      <td>S302</td>\n",
       "      <td>GrGGrGGsrArNLPL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18450</td>\n",
       "      <td>NuaK1</td>\n",
       "      <td>O60285</td>\n",
       "      <td>MYPT1</td>\n",
       "      <td>O14974</td>\n",
       "      <td>PPP1R12A</td>\n",
       "      <td>S910</td>\n",
       "      <td>sLLGRsGsysyLEER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18451</td>\n",
       "      <td>NuaK1</td>\n",
       "      <td>O60285</td>\n",
       "      <td>LATS1</td>\n",
       "      <td>O95835</td>\n",
       "      <td>LATS1</td>\n",
       "      <td>S464</td>\n",
       "      <td>NIPVRsNsFNNPLGN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18452</td>\n",
       "      <td>ULK2</td>\n",
       "      <td>Q8IYT8</td>\n",
       "      <td>SEC16A</td>\n",
       "      <td>O15027</td>\n",
       "      <td>SEC16A</td>\n",
       "      <td>S846</td>\n",
       "      <td>LAQPINFsVSLSNSH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18453</td>\n",
       "      <td>ULK2</td>\n",
       "      <td>Q8IYT8</td>\n",
       "      <td>DENND3</td>\n",
       "      <td>A2RUS2</td>\n",
       "      <td>DENND3</td>\n",
       "      <td>S472</td>\n",
       "      <td>THRRMVVsMPNLQDI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18454</td>\n",
       "      <td>ULK2</td>\n",
       "      <td>Q8IYT8</td>\n",
       "      <td>DENND3</td>\n",
       "      <td>A2RUS2</td>\n",
       "      <td>DENND3</td>\n",
       "      <td>S490</td>\n",
       "      <td>ELAPRNSsLRLtDTA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10981 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      KINASE_GENE_NAME KIN_UNIPROT_ID SUBSTRATE_NAME SUB_UNIPROT_ID  \\\n",
       "6                  HRI         Q9BQI3     eIF2-alpha         P05198   \n",
       "7                  HRI         Q9BQI3     eIF2-alpha         P05198   \n",
       "10                PKCD         Q05655          HDAC5         Q9UQL6   \n",
       "11                PKCD         Q05655     PTPRA iso2       P18433-2   \n",
       "12                PKCD         Q05655        hnRNP K         P61978   \n",
       "...                ...            ...            ...            ...   \n",
       "18450            NuaK1         O60285          MYPT1         O14974   \n",
       "18451            NuaK1         O60285          LATS1         O95835   \n",
       "18452             ULK2         Q8IYT8         SEC16A         O15027   \n",
       "18453             ULK2         Q8IYT8         DENND3         A2RUS2   \n",
       "18454             ULK2         Q8IYT8         DENND3         A2RUS2   \n",
       "\n",
       "      SUB_GENE_NAME SUB_MOD_RSD        SITE_7_AA  \n",
       "6            EIF2S1         S52  MILLsELsRRRIRsI  \n",
       "7            EIF2S1         S49  IEGMILLsELsRRRI  \n",
       "10            HDAC5        S259  FPLRkTAsEPNLKVR  \n",
       "11            PTPRA        S204  PLLARSPsTNRKYPP  \n",
       "12           HNRNPK        S302  GrGGrGGsrArNLPL  \n",
       "...             ...         ...              ...  \n",
       "18450      PPP1R12A        S910  sLLGRsGsysyLEER  \n",
       "18451         LATS1        S464  NIPVRsNsFNNPLGN  \n",
       "18452        SEC16A        S846  LAQPINFsVSLSNSH  \n",
       "18453        DENND3        S472  THRRMVVsMPNLQDI  \n",
       "18454        DENND3        S490  ELAPRNSsLRLtDTA  \n",
       "\n",
       "[10981 rows x 7 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Renaming columns of the dataframe for more clarity and to meet syntax requirements for SQLite and SQLAlchemy.\n",
    "df = df.rename(index=str, columns = {'KINASE': 'KINASE_GENE_NAME', 'KIN_ACC_ID': 'KIN_UNIPROT_ID', 'SUBSTRATE': 'SUBSTRATE_NAME',\n",
    "                                     'SUB_ACC_ID': 'SUB_UNIPROT_ID', 'SUB_GENE': 'SUB_GENE_NAME', 'SITE_+/-7_AA': 'SITE_7_AA'}) \n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The neighbouring amino acid sequence shows the 7 neighbouring amino acids either side of the amino acid residue that has been phosphorylated by the Kinase. However, the format that PhosphositePlus presents the information isn't the clearest with multiple lowercase amino acids in the same string that aren't the residue being phosphorylated. So we need to clean the column to make them look the same way."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Putting the Amino acid sequence information into an object\n",
    "AA_Sequences = df['SITE_7_AA'].values\n",
    "\n",
    "#Make a function which takes a string and capitalizes the 8th character in a string and makes all of the other characters in\n",
    "#lower case.\n",
    "def capitalize_8th(string):\n",
    "    return string[:7].lower() + string[7:].capitalize()\n",
    "\n",
    "#Set up an empty list to collect all of the cleaned sequences\n",
    "CleanedSequences = []\n",
    "\n",
    "#Running a for loop to clean the amino acid sequences\n",
    "for sequence in AA_Sequences: #for every sequence in the amino acid sequence\n",
    "    sequence = str(sequence) #convert the sequence into a string\n",
    "    sequence = capitalize_8th(sequence) #pass the string through our created function \n",
    "    sequence = sequence.swapcase() #swap the cases of the strings to the opposite i.e. all lowercase characters become uppercase\n",
    "                                   #and vice versa\n",
    "    CleanedSequences.append(sequence) #append the edited sequence to the empty list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>KINASE_GENE_NAME</th>\n",
       "      <th>KIN_UNIPROT_ID</th>\n",
       "      <th>SUBSTRATE_NAME</th>\n",
       "      <th>SUB_UNIPROT_ID</th>\n",
       "      <th>SUB_GENE_NAME</th>\n",
       "      <th>SUB_MOD_RSD</th>\n",
       "      <th>SITE_7_AA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>HRI</td>\n",
       "      <td>Q9BQI3</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>S52</td>\n",
       "      <td>MILLSELsRRRIRSI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>HRI</td>\n",
       "      <td>Q9BQI3</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>S49</td>\n",
       "      <td>IEGMILLsELSRRRI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>PKCD</td>\n",
       "      <td>Q05655</td>\n",
       "      <td>HDAC5</td>\n",
       "      <td>Q9UQL6</td>\n",
       "      <td>HDAC5</td>\n",
       "      <td>S259</td>\n",
       "      <td>FPLRKTAsEPNLKVR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>PKCD</td>\n",
       "      <td>Q05655</td>\n",
       "      <td>PTPRA iso2</td>\n",
       "      <td>P18433-2</td>\n",
       "      <td>PTPRA</td>\n",
       "      <td>S204</td>\n",
       "      <td>PLLARSPsTNRKYPP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>PKCD</td>\n",
       "      <td>Q05655</td>\n",
       "      <td>hnRNP K</td>\n",
       "      <td>P61978</td>\n",
       "      <td>HNRNPK</td>\n",
       "      <td>S302</td>\n",
       "      <td>GRGGRGGsRARNLPL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18450</td>\n",
       "      <td>NuaK1</td>\n",
       "      <td>O60285</td>\n",
       "      <td>MYPT1</td>\n",
       "      <td>O14974</td>\n",
       "      <td>PPP1R12A</td>\n",
       "      <td>S910</td>\n",
       "      <td>SLLGRSGsYSYLEER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18451</td>\n",
       "      <td>NuaK1</td>\n",
       "      <td>O60285</td>\n",
       "      <td>LATS1</td>\n",
       "      <td>O95835</td>\n",
       "      <td>LATS1</td>\n",
       "      <td>S464</td>\n",
       "      <td>NIPVRSNsFNNPLGN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18452</td>\n",
       "      <td>ULK2</td>\n",
       "      <td>Q8IYT8</td>\n",
       "      <td>SEC16A</td>\n",
       "      <td>O15027</td>\n",
       "      <td>SEC16A</td>\n",
       "      <td>S846</td>\n",
       "      <td>LAQPINFsVSLSNSH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18453</td>\n",
       "      <td>ULK2</td>\n",
       "      <td>Q8IYT8</td>\n",
       "      <td>DENND3</td>\n",
       "      <td>A2RUS2</td>\n",
       "      <td>DENND3</td>\n",
       "      <td>S472</td>\n",
       "      <td>THRRMVVsMPNLQDI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18454</td>\n",
       "      <td>ULK2</td>\n",
       "      <td>Q8IYT8</td>\n",
       "      <td>DENND3</td>\n",
       "      <td>A2RUS2</td>\n",
       "      <td>DENND3</td>\n",
       "      <td>S490</td>\n",
       "      <td>ELAPRNSsLRLTDTA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10981 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      KINASE_GENE_NAME KIN_UNIPROT_ID SUBSTRATE_NAME SUB_UNIPROT_ID  \\\n",
       "6                  HRI         Q9BQI3     eIF2-alpha         P05198   \n",
       "7                  HRI         Q9BQI3     eIF2-alpha         P05198   \n",
       "10                PKCD         Q05655          HDAC5         Q9UQL6   \n",
       "11                PKCD         Q05655     PTPRA iso2       P18433-2   \n",
       "12                PKCD         Q05655        hnRNP K         P61978   \n",
       "...                ...            ...            ...            ...   \n",
       "18450            NuaK1         O60285          MYPT1         O14974   \n",
       "18451            NuaK1         O60285          LATS1         O95835   \n",
       "18452             ULK2         Q8IYT8         SEC16A         O15027   \n",
       "18453             ULK2         Q8IYT8         DENND3         A2RUS2   \n",
       "18454             ULK2         Q8IYT8         DENND3         A2RUS2   \n",
       "\n",
       "      SUB_GENE_NAME SUB_MOD_RSD        SITE_7_AA  \n",
       "6            EIF2S1         S52  MILLSELsRRRIRSI  \n",
       "7            EIF2S1         S49  IEGMILLsELSRRRI  \n",
       "10            HDAC5        S259  FPLRKTAsEPNLKVR  \n",
       "11            PTPRA        S204  PLLARSPsTNRKYPP  \n",
       "12           HNRNPK        S302  GRGGRGGsRARNLPL  \n",
       "...             ...         ...              ...  \n",
       "18450      PPP1R12A        S910  SLLGRSGsYSYLEER  \n",
       "18451         LATS1        S464  NIPVRSNsFNNPLGN  \n",
       "18452        SEC16A        S846  LAQPINFsVSLSNSH  \n",
       "18453        DENND3        S472  THRRMVVsMPNLQDI  \n",
       "18454        DENND3        S490  ELAPRNSsLRLTDTA  \n",
       "\n",
       "[10981 rows x 7 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Dropping the old amino acid sequence column in the dataframe\n",
    "df = df.drop(columns=['SITE_7_AA'])\n",
    "\n",
    "#Inserting the new column with the cleaned data\n",
    "df.insert(6, 'SITE_7_AA', CleanedSequences)\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that the Kinase_Substrate_Dataset has been filtered and cleaned to our specifications we now need to incorproate any further required information from the Phosphorylation_site_dataset. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>PROTEIN</th>\n",
       "      <th>ACC_ID</th>\n",
       "      <th>HU_CHR_LOC</th>\n",
       "      <th>MOD_RSD</th>\n",
       "      <th>SITE_GRP_ID</th>\n",
       "      <th>ORGANISM</th>\n",
       "      <th>MW_kD</th>\n",
       "      <th>DOMAIN</th>\n",
       "      <th>SITE_+/-7_AA</th>\n",
       "      <th>LT_LIT</th>\n",
       "      <th>MS_LIT</th>\n",
       "      <th>MS_CST</th>\n",
       "      <th>CST_CAT#</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>1110035H17Rik</td>\n",
       "      <td>1110035H17Rik</td>\n",
       "      <td>Q9CTA4</td>\n",
       "      <td>7|7</td>\n",
       "      <td>S10-p</td>\n",
       "      <td>7231581</td>\n",
       "      <td>mouse</td>\n",
       "      <td>24.31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RPPPGSRstVAQSPP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>1110035H17Rik</td>\n",
       "      <td>1110035H17Rik</td>\n",
       "      <td>Q9CTA4</td>\n",
       "      <td>7|7</td>\n",
       "      <td>T11-p</td>\n",
       "      <td>7231583</td>\n",
       "      <td>mouse</td>\n",
       "      <td>24.31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PPPGSRstVAQSPPQ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>YWHAB</td>\n",
       "      <td>14-3-3 beta</td>\n",
       "      <td>P31946</td>\n",
       "      <td>20q13.12</td>\n",
       "      <td>T2-p</td>\n",
       "      <td>15718712</td>\n",
       "      <td>human</td>\n",
       "      <td>28.08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>______MtMDksELV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Ywhab</td>\n",
       "      <td>14-3-3 beta</td>\n",
       "      <td>Q9CQV8</td>\n",
       "      <td>2|2 H3</td>\n",
       "      <td>T2-p</td>\n",
       "      <td>15718712</td>\n",
       "      <td>mouse</td>\n",
       "      <td>28.09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>______MtMDksELV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>YWHAB</td>\n",
       "      <td>14-3-3 beta</td>\n",
       "      <td>P31946</td>\n",
       "      <td>20q13.12</td>\n",
       "      <td>S6-p</td>\n",
       "      <td>15718709</td>\n",
       "      <td>human</td>\n",
       "      <td>28.08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__MtMDksELVQkAk</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            GENE        PROTEIN  ACC_ID HU_CHR_LOC MOD_RSD  SITE_GRP_ID  \\\n",
       "0  1110035H17Rik  1110035H17Rik  Q9CTA4        7|7   S10-p      7231581   \n",
       "1  1110035H17Rik  1110035H17Rik  Q9CTA4        7|7   T11-p      7231583   \n",
       "2          YWHAB    14-3-3 beta  P31946   20q13.12    T2-p     15718712   \n",
       "3          Ywhab    14-3-3 beta  Q9CQV8     2|2 H3    T2-p     15718712   \n",
       "4          YWHAB    14-3-3 beta  P31946   20q13.12    S6-p     15718709   \n",
       "\n",
       "  ORGANISM  MW_kD DOMAIN     SITE_+/-7_AA  LT_LIT  MS_LIT  MS_CST CST_CAT#  \n",
       "0    mouse  24.31    NaN  RPPPGSRstVAQSPP     NaN     1.0     NaN      NaN  \n",
       "1    mouse  24.31    NaN  PPPGSRstVAQSPPQ     NaN     1.0     NaN      NaN  \n",
       "2    human  28.08    NaN  ______MtMDksELV     NaN     3.0     1.0      NaN  \n",
       "3    mouse  28.09    NaN  ______MtMDksELV     NaN     2.0     NaN      NaN  \n",
       "4    human  28.08    NaN  __MtMDksELVQkAk     NaN     8.0     NaN      NaN  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Load in the Phosphorylation_site_dataset into pandas and check to make sure it has loaded in properly.\n",
    "df1 = pd.read_csv('Phosphorylation_site_dataset.csv')\n",
    "df1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>PROTEIN</th>\n",
       "      <th>ACC_ID</th>\n",
       "      <th>HU_CHR_LOC</th>\n",
       "      <th>MOD_RSD</th>\n",
       "      <th>SITE_GRP_ID</th>\n",
       "      <th>ORGANISM</th>\n",
       "      <th>MW_kD</th>\n",
       "      <th>DOMAIN</th>\n",
       "      <th>SITE_+/-7_AA</th>\n",
       "      <th>LT_LIT</th>\n",
       "      <th>MS_LIT</th>\n",
       "      <th>MS_CST</th>\n",
       "      <th>CST_CAT#</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>YWHAB</td>\n",
       "      <td>14-3-3 beta</td>\n",
       "      <td>P31946</td>\n",
       "      <td>20q13.12</td>\n",
       "      <td>T2-p</td>\n",
       "      <td>15718712</td>\n",
       "      <td>human</td>\n",
       "      <td>28.08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>______MtMDksELV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>YWHAB</td>\n",
       "      <td>14-3-3 beta</td>\n",
       "      <td>P31946</td>\n",
       "      <td>20q13.12</td>\n",
       "      <td>S6-p</td>\n",
       "      <td>15718709</td>\n",
       "      <td>human</td>\n",
       "      <td>28.08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>__MtMDksELVQkAk</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>YWHAB</td>\n",
       "      <td>14-3-3 beta</td>\n",
       "      <td>P31946</td>\n",
       "      <td>20q13.12</td>\n",
       "      <td>Y21-p</td>\n",
       "      <td>3426383</td>\n",
       "      <td>human</td>\n",
       "      <td>28.08</td>\n",
       "      <td>14-3-3</td>\n",
       "      <td>LAEQAERyDDMAAAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>YWHAB</td>\n",
       "      <td>14-3-3 beta</td>\n",
       "      <td>P31946</td>\n",
       "      <td>20q13.12</td>\n",
       "      <td>T32-p</td>\n",
       "      <td>23077803</td>\n",
       "      <td>human</td>\n",
       "      <td>28.08</td>\n",
       "      <td>14-3-3</td>\n",
       "      <td>AAAMkAVtEQGHELs</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>YWHAB</td>\n",
       "      <td>14-3-3 beta</td>\n",
       "      <td>P31946</td>\n",
       "      <td>20q13.12</td>\n",
       "      <td>S39-p</td>\n",
       "      <td>27442700</td>\n",
       "      <td>human</td>\n",
       "      <td>28.08</td>\n",
       "      <td>14-3-3</td>\n",
       "      <td>tEQGHELsNEERNLL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    GENE      PROTEIN  ACC_ID HU_CHR_LOC MOD_RSD  SITE_GRP_ID ORGANISM  MW_kD  \\\n",
       "2  YWHAB  14-3-3 beta  P31946   20q13.12    T2-p     15718712    human  28.08   \n",
       "4  YWHAB  14-3-3 beta  P31946   20q13.12    S6-p     15718709    human  28.08   \n",
       "6  YWHAB  14-3-3 beta  P31946   20q13.12   Y21-p      3426383    human  28.08   \n",
       "8  YWHAB  14-3-3 beta  P31946   20q13.12   T32-p     23077803    human  28.08   \n",
       "9  YWHAB  14-3-3 beta  P31946   20q13.12   S39-p     27442700    human  28.08   \n",
       "\n",
       "   DOMAIN     SITE_+/-7_AA  LT_LIT  MS_LIT  MS_CST CST_CAT#  \n",
       "2     NaN  ______MtMDksELV     NaN     3.0     1.0      NaN  \n",
       "4     NaN  __MtMDksELVQkAk     NaN     8.0     NaN      NaN  \n",
       "6  14-3-3  LAEQAERyDDMAAAM     NaN     NaN     4.0      NaN  \n",
       "8  14-3-3  AAAMkAVtEQGHELs     NaN     NaN     1.0      NaN  \n",
       "9  14-3-3  tEQGHELsNEERNLL     NaN     4.0     NaN      NaN  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Similar to the Kinase_substrate_dataset we see that there is data on organisms that aren't human. So we filter the dataframe \n",
    "#in the same way to only have rows where the organism is human and check to make sure this has been done.\n",
    "df1 = df1[df1.ORGANISM == 'human']\n",
    "df1.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Looking at the data in the Phosphorylation_site_dataset, the only column of interest we'd want to include in our final PhosphoSite table is the Human Chromosome Location. To incorporate Human Chromosome Location we first removed all of the rwos except for the Human Chromosome Location and a column that was shared between the two datasets i.e. ACC_ID which is equivalent to SUB_UNIPROT_ID in our cleaned Kinase_substrate_dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ACC_ID</th>\n",
       "      <th>HU_CHR_LOC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>P31946</td>\n",
       "      <td>20q13.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>P31946</td>\n",
       "      <td>20q13.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>P31946</td>\n",
       "      <td>20q13.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>P31946</td>\n",
       "      <td>20q13.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>P31946</td>\n",
       "      <td>20q13.12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ACC_ID HU_CHR_LOC\n",
       "2  P31946   20q13.12\n",
       "4  P31946   20q13.12\n",
       "6  P31946   20q13.12\n",
       "8  P31946   20q13.12\n",
       "9  P31946   20q13.12"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Dropping all columns in the Phosphorylation_site_dataset except ACC_ID and HU_CHR_LOC\n",
    "\n",
    "df1 = df1.drop(columns=['GENE','PROTEIN', 'MOD_RSD', 'SITE_GRP_ID','ORGANISM', 'MW_kD', 'DOMAIN', 'SITE_+/-7_AA', 'LT_LIT',\n",
    "                        'MS_LIT', 'MS_CST', 'CST_CAT#'])\n",
    "df1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUB_UNIPROT_ID</th>\n",
       "      <th>CHR_LOC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>P31946</td>\n",
       "      <td>20q13.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>P31946</td>\n",
       "      <td>20q13.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>P31946</td>\n",
       "      <td>20q13.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>P31946</td>\n",
       "      <td>20q13.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>P31946</td>\n",
       "      <td>20q13.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>371197</td>\n",
       "      <td>Q8IYH5</td>\n",
       "      <td>1p31.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>371198</td>\n",
       "      <td>Q8IYH5</td>\n",
       "      <td>1p31.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>371200</td>\n",
       "      <td>Q8IYH5</td>\n",
       "      <td>1p31.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>371201</td>\n",
       "      <td>Q8IYH5</td>\n",
       "      <td>1p31.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>371202</td>\n",
       "      <td>Q8IYH5</td>\n",
       "      <td>1p31.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>239530 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       SUB_UNIPROT_ID   CHR_LOC\n",
       "2              P31946  20q13.12\n",
       "4              P31946  20q13.12\n",
       "6              P31946  20q13.12\n",
       "8              P31946  20q13.12\n",
       "9              P31946  20q13.12\n",
       "...               ...       ...\n",
       "371197         Q8IYH5    1p31.1\n",
       "371198         Q8IYH5    1p31.1\n",
       "371200         Q8IYH5    1p31.1\n",
       "371201         Q8IYH5    1p31.1\n",
       "371202         Q8IYH5    1p31.1\n",
       "\n",
       "[239530 rows x 2 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#We re-name the ACC_ID column to match our other dataset in order to actually carry out the next merge step.\n",
    "df1 = df1.rename(index=str, columns = {'ACC_ID': 'SUB_UNIPROT_ID', 'HU_CHR_LOC': 'CHR_LOC'}) \n",
    "df1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As there is a massive difference in the number of rows between our two dataframes (10981 vs 239530 rows), in order for the dataframe merge to be done correctly the difference between the two needs to be reduced significantly. As df1 has many duplicates to make up its number of rows we go about dropping the duplicates in order to reduce the row numbers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUB_UNIPROT_ID</th>\n",
       "      <th>CHR_LOC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>P31946</td>\n",
       "      <td>20q13.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>49</td>\n",
       "      <td>P62258</td>\n",
       "      <td>17p13.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>85</td>\n",
       "      <td>Q04917</td>\n",
       "      <td>22q12.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>115</td>\n",
       "      <td>P61981</td>\n",
       "      <td>7q11.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>161</td>\n",
       "      <td>P31947</td>\n",
       "      <td>1p36.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>371037</td>\n",
       "      <td>Q9C0D3</td>\n",
       "      <td>1p32.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>371042</td>\n",
       "      <td>Q7Z7L7</td>\n",
       "      <td>9q34.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>371050</td>\n",
       "      <td>Q15942</td>\n",
       "      <td>7q34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>371107</td>\n",
       "      <td>O43149</td>\n",
       "      <td>17p13.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>371178</td>\n",
       "      <td>Q8IYH5</td>\n",
       "      <td>1p31.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20101 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       SUB_UNIPROT_ID   CHR_LOC\n",
       "2              P31946  20q13.12\n",
       "49             P62258   17p13.3\n",
       "85             Q04917   22q12.3\n",
       "115            P61981   7q11.23\n",
       "161            P31947   1p36.11\n",
       "...               ...       ...\n",
       "371037         Q9C0D3    1p32.3\n",
       "371042         Q7Z7L7   9q34.11\n",
       "371050         Q15942      7q34\n",
       "371107         O43149   17p13.2\n",
       "371178         Q8IYH5    1p31.1\n",
       "\n",
       "[20101 rows x 2 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Dropping duplicates in Dataframe1.\n",
    "df1 = df1.drop_duplicates(keep = 'first', inplace=False)\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>KINASE_GENE_NAME</th>\n",
       "      <th>KIN_UNIPROT_ID</th>\n",
       "      <th>SUBSTRATE_NAME</th>\n",
       "      <th>SUB_UNIPROT_ID</th>\n",
       "      <th>SUB_GENE_NAME</th>\n",
       "      <th>SUB_MOD_RSD</th>\n",
       "      <th>SITE_7_AA</th>\n",
       "      <th>CHR_LOC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>HRI</td>\n",
       "      <td>Q9BQI3</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>S52</td>\n",
       "      <td>MILLSELsRRRIRSI</td>\n",
       "      <td>14q23.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>HRI</td>\n",
       "      <td>Q9BQI3</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>S49</td>\n",
       "      <td>IEGMILLsELSRRRI</td>\n",
       "      <td>14q23.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>PKR</td>\n",
       "      <td>P19525</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>S49</td>\n",
       "      <td>IEGMILLsELSRRRI</td>\n",
       "      <td>14q23.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>PKR</td>\n",
       "      <td>P19525</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>S52</td>\n",
       "      <td>MILLSELsRRRIRSI</td>\n",
       "      <td>14q23.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>P38A</td>\n",
       "      <td>Q16539</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>S52</td>\n",
       "      <td>MILLSELsRRRIRSI</td>\n",
       "      <td>14q23.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10976</td>\n",
       "      <td>ZAK</td>\n",
       "      <td>Q9NYL2</td>\n",
       "      <td>ZAK</td>\n",
       "      <td>Q9NYL2</td>\n",
       "      <td>MAP3K20</td>\n",
       "      <td>T162</td>\n",
       "      <td>SRFHNHTtHMSLVGT</td>\n",
       "      <td>2q31.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10977</td>\n",
       "      <td>ZAK</td>\n",
       "      <td>Q9NYL2</td>\n",
       "      <td>ZAK</td>\n",
       "      <td>Q9NYL2</td>\n",
       "      <td>MAP3K20</td>\n",
       "      <td>S165</td>\n",
       "      <td>HNHTTHMsLVGTFPW</td>\n",
       "      <td>2q31.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10978</td>\n",
       "      <td>ZAK</td>\n",
       "      <td>Q9NYL2</td>\n",
       "      <td>ZAK</td>\n",
       "      <td>Q9NYL2</td>\n",
       "      <td>MAP3K20</td>\n",
       "      <td>T161</td>\n",
       "      <td>ASRFHNHtTHMSLVG</td>\n",
       "      <td>2q31.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10979</td>\n",
       "      <td>MEKK4</td>\n",
       "      <td>Q9Y6R4</td>\n",
       "      <td>MEKK4</td>\n",
       "      <td>Q9Y6R4</td>\n",
       "      <td>MAP3K4</td>\n",
       "      <td>T1494</td>\n",
       "      <td>KLKNNAQtMPGEVNS</td>\n",
       "      <td>6q26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10980</td>\n",
       "      <td>NuaK1</td>\n",
       "      <td>O60285</td>\n",
       "      <td>CASP6</td>\n",
       "      <td>P55212</td>\n",
       "      <td>CASP6</td>\n",
       "      <td>S257</td>\n",
       "      <td>TLVNRKVsQRRVDFC</td>\n",
       "      <td>4q25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10981 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      KINASE_GENE_NAME KIN_UNIPROT_ID SUBSTRATE_NAME SUB_UNIPROT_ID  \\\n",
       "0                  HRI         Q9BQI3     eIF2-alpha         P05198   \n",
       "1                  HRI         Q9BQI3     eIF2-alpha         P05198   \n",
       "2                  PKR         P19525     eIF2-alpha         P05198   \n",
       "3                  PKR         P19525     eIF2-alpha         P05198   \n",
       "4                 P38A         Q16539     eIF2-alpha         P05198   \n",
       "...                ...            ...            ...            ...   \n",
       "10976              ZAK         Q9NYL2            ZAK         Q9NYL2   \n",
       "10977              ZAK         Q9NYL2            ZAK         Q9NYL2   \n",
       "10978              ZAK         Q9NYL2            ZAK         Q9NYL2   \n",
       "10979            MEKK4         Q9Y6R4          MEKK4         Q9Y6R4   \n",
       "10980            NuaK1         O60285          CASP6         P55212   \n",
       "\n",
       "      SUB_GENE_NAME SUB_MOD_RSD        SITE_7_AA  CHR_LOC  \n",
       "0            EIF2S1         S52  MILLSELsRRRIRSI  14q23.3  \n",
       "1            EIF2S1         S49  IEGMILLsELSRRRI  14q23.3  \n",
       "2            EIF2S1         S49  IEGMILLsELSRRRI  14q23.3  \n",
       "3            EIF2S1         S52  MILLSELsRRRIRSI  14q23.3  \n",
       "4            EIF2S1         S52  MILLSELsRRRIRSI  14q23.3  \n",
       "...             ...         ...              ...      ...  \n",
       "10976       MAP3K20        T162  SRFHNHTtHMSLVGT   2q31.1  \n",
       "10977       MAP3K20        S165  HNHTTHMsLVGTFPW   2q31.1  \n",
       "10978       MAP3K20        T161  ASRFHNHtTHMSLVG   2q31.1  \n",
       "10979        MAP3K4       T1494  KLKNNAQtMPGEVNS     6q26  \n",
       "10980         CASP6        S257  TLVNRKVsQRRVDFC     4q25  \n",
       "\n",
       "[10981 rows x 8 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Now that the rows are reduced, we merge the dataframes together by the SUB_UNIPROT_ID so that Human Chromosome Location is now\n",
    "#included to our cleaned table.\n",
    "df = pd.merge(df,df1[['SUB_UNIPROT_ID','CHR_LOC']],on='SUB_UNIPROT_ID', how='inner')\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This table normally on its own would be more than enough for our database. However, as part of the data visualisation part of the web application, we would need to cross reference our phosphosite table of our database in order to carry out some visualisations. As a some of the substrates in example phosphoproteomic data file contain UniProt entry names in addition to gene names, we thought it would also be useful to include a column containing all of the corresponding UniProt entry names to each of the substrates. \n",
    "\n",
    "To do this we can run our existing UniProt IDs through the UniProt API where they can return our corresponding entry names. As our current df has almost 11k rows which would take the UniProt API a long time to process through. We can reduce the processing time by dropping duplicate UniProt IDs from our dataframe and re-merging the corresponding entries to our dataframe at a later point."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUB_UNIPROT_ID</th>\n",
       "      <th>SUB_GENE_NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>Q9UQL6</td>\n",
       "      <td>HDAC5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>30</td>\n",
       "      <td>P18433-2</td>\n",
       "      <td>PTPRA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>33</td>\n",
       "      <td>P61978</td>\n",
       "      <td>HNRNPK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>48</td>\n",
       "      <td>Q9UQ13</td>\n",
       "      <td>SHOC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10963</td>\n",
       "      <td>Q9UJV9</td>\n",
       "      <td>DDX41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10964</td>\n",
       "      <td>Q9NWZ3</td>\n",
       "      <td>IRAK4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10976</td>\n",
       "      <td>Q9NYL2</td>\n",
       "      <td>MAP3K20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10979</td>\n",
       "      <td>Q9Y6R4</td>\n",
       "      <td>MAP3K4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10980</td>\n",
       "      <td>P55212</td>\n",
       "      <td>CASP6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2531 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      SUB_UNIPROT_ID SUB_GENE_NAME\n",
       "0             P05198        EIF2S1\n",
       "9             Q9UQL6         HDAC5\n",
       "30          P18433-2         PTPRA\n",
       "33            P61978        HNRNPK\n",
       "48            Q9UQ13         SHOC2\n",
       "...              ...           ...\n",
       "10963         Q9UJV9         DDX41\n",
       "10964         Q9NWZ3         IRAK4\n",
       "10976         Q9NYL2       MAP3K20\n",
       "10979         Q9Y6R4        MAP3K4\n",
       "10980         P55212         CASP6\n",
       "\n",
       "[2531 rows x 2 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#We drop all of our columns except for SUB_UNIPROT_ID and SUB_GENE_NAME and place it into another dataframe object then we can \n",
    "#drop any duplicates.\n",
    "df2 = df.drop(columns=['KINASE_GENE_NAME','KIN_UNIPROT_ID', 'SUBSTRATE_NAME', 'SUB_MOD_RSD','SITE_7_AA', 'CHR_LOC'])\n",
    "df2 = df2.drop_duplicates(keep = 'first', inplace=False)\n",
    "\n",
    "#Reduces the number of rows from 10981 rows and 2531 rows.\n",
    "df2 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Now we put all of our UniProt IDs into another object in order to feed through the UniProt API.\n",
    "UniProt_IDs = df2['SUB_UNIPROT_ID'].values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When we tried to run our UniProt IDs through the UniProt API using a For loop, we wound up getting errors and after investigating the which UniProt IDs caused these errors fell into three categories:\n",
    "\n",
    "1) UniProt IDs for isoforms of a protein that are labelled a '-' eg. P18433-2\n",
    "\n",
    "2) UniProt IDs that contained a specific variation of a protein that are labelled with '_VAR_' e.g. O60504_VAR_055019\n",
    "\n",
    "3) IDs that don't really fall into either of the two above categories and don't actually have a page on UniProt when searched manually e.g. AAC50053\n",
    "\n",
    "So to fix the errors caused by 1 and 2 we used a regex to remove all of the characters from a '-' or '_' onwards. As we were unable to find a corresponding UniProt entries for IDs that fell into category 3 we simply assigned them a blank entry."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create an empty list to contain all of our UniProt Entries.\n",
    "EntryList = []\n",
    "\n",
    "for ID in UniProt_IDs: #for every ID in our object containing all of our UniProt IDs\n",
    "    ID = str(ID) #convert the ID into a string\n",
    "    ID = re.sub('([-,_].*)$', '', ID) #apply a regex to clean ID string in categories 1 and 2.\n",
    "    try: #pass cleaned IDs through UniProt API to get a table containing ID and corresponding UniProt Entry name read by pandas                              \n",
    "        data = pd.read_csv(\"https://www.uniprot.org/uniprot/?query=\"+ID+\"&sort=score&columns=id,entry%20name&format=tab\", \n",
    "                           sep=\"\\t\") \n",
    "        entry = data[\"Entry name\"][0] #we then store the first value in the Entry name column of the table as this corresponds\n",
    "                                      #to our specific ID.\n",
    "        EntryList.append(entry) #We then append that entry name into our empty list\n",
    "    except:\n",
    "        EntryList.append('-') #In the case, of IDs in category 3 we would just append a '-' to represent a blank."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUB_ENTRY_NAME</th>\n",
       "      <th>SUB_UNIPROT_ID</th>\n",
       "      <th>SUB_GENE_NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>IF2A_HUMAN</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>HDAC5_HUMAN</td>\n",
       "      <td>Q9UQL6</td>\n",
       "      <td>HDAC5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>30</td>\n",
       "      <td>PTPRA_HUMAN</td>\n",
       "      <td>P18433-2</td>\n",
       "      <td>PTPRA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>33</td>\n",
       "      <td>HNRPK_HUMAN</td>\n",
       "      <td>P61978</td>\n",
       "      <td>HNRNPK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>48</td>\n",
       "      <td>SHOC2_HUMAN</td>\n",
       "      <td>Q9UQ13</td>\n",
       "      <td>SHOC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10963</td>\n",
       "      <td>DDX41_HUMAN</td>\n",
       "      <td>Q9UJV9</td>\n",
       "      <td>DDX41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10964</td>\n",
       "      <td>IRAK4_HUMAN</td>\n",
       "      <td>Q9NWZ3</td>\n",
       "      <td>IRAK4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10976</td>\n",
       "      <td>M3K20_HUMAN</td>\n",
       "      <td>Q9NYL2</td>\n",
       "      <td>MAP3K20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10979</td>\n",
       "      <td>M3K4_HUMAN</td>\n",
       "      <td>Q9Y6R4</td>\n",
       "      <td>MAP3K4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10980</td>\n",
       "      <td>CASP6_HUMAN</td>\n",
       "      <td>P55212</td>\n",
       "      <td>CASP6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2531 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      SUB_ENTRY_NAME SUB_UNIPROT_ID SUB_GENE_NAME\n",
       "0         IF2A_HUMAN         P05198        EIF2S1\n",
       "9        HDAC5_HUMAN         Q9UQL6         HDAC5\n",
       "30       PTPRA_HUMAN       P18433-2         PTPRA\n",
       "33       HNRPK_HUMAN         P61978        HNRNPK\n",
       "48       SHOC2_HUMAN         Q9UQ13         SHOC2\n",
       "...              ...            ...           ...\n",
       "10963    DDX41_HUMAN         Q9UJV9         DDX41\n",
       "10964    IRAK4_HUMAN         Q9NWZ3         IRAK4\n",
       "10976    M3K20_HUMAN         Q9NYL2       MAP3K20\n",
       "10979     M3K4_HUMAN         Q9Y6R4        MAP3K4\n",
       "10980    CASP6_HUMAN         P55212         CASP6\n",
       "\n",
       "[2531 rows x 3 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Now we insert our list of UniProt entries to our second dataframe...\n",
    "df2.insert(0, 'SUB_ENTRY_NAME', EntryList)\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>KINASE_GENE_NAME</th>\n",
       "      <th>KIN_UNIPROT_ID</th>\n",
       "      <th>SUBSTRATE_NAME</th>\n",
       "      <th>SUB_UNIPROT_ID</th>\n",
       "      <th>SUB_GENE_NAME</th>\n",
       "      <th>SUB_MOD_RSD</th>\n",
       "      <th>SITE_7_AA</th>\n",
       "      <th>CHR_LOC</th>\n",
       "      <th>SUB_ENTRY_NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>HRI</td>\n",
       "      <td>Q9BQI3</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>S52</td>\n",
       "      <td>MILLSELsRRRIRSI</td>\n",
       "      <td>14q23.3</td>\n",
       "      <td>IF2A_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>HRI</td>\n",
       "      <td>Q9BQI3</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>S49</td>\n",
       "      <td>IEGMILLsELSRRRI</td>\n",
       "      <td>14q23.3</td>\n",
       "      <td>IF2A_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>PKR</td>\n",
       "      <td>P19525</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>S49</td>\n",
       "      <td>IEGMILLsELSRRRI</td>\n",
       "      <td>14q23.3</td>\n",
       "      <td>IF2A_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>PKR</td>\n",
       "      <td>P19525</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>S52</td>\n",
       "      <td>MILLSELsRRRIRSI</td>\n",
       "      <td>14q23.3</td>\n",
       "      <td>IF2A_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>P38A</td>\n",
       "      <td>Q16539</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>S52</td>\n",
       "      <td>MILLSELsRRRIRSI</td>\n",
       "      <td>14q23.3</td>\n",
       "      <td>IF2A_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10976</td>\n",
       "      <td>ZAK</td>\n",
       "      <td>Q9NYL2</td>\n",
       "      <td>ZAK</td>\n",
       "      <td>Q9NYL2</td>\n",
       "      <td>MAP3K20</td>\n",
       "      <td>T162</td>\n",
       "      <td>SRFHNHTtHMSLVGT</td>\n",
       "      <td>2q31.1</td>\n",
       "      <td>M3K20_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10977</td>\n",
       "      <td>ZAK</td>\n",
       "      <td>Q9NYL2</td>\n",
       "      <td>ZAK</td>\n",
       "      <td>Q9NYL2</td>\n",
       "      <td>MAP3K20</td>\n",
       "      <td>S165</td>\n",
       "      <td>HNHTTHMsLVGTFPW</td>\n",
       "      <td>2q31.1</td>\n",
       "      <td>M3K20_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10978</td>\n",
       "      <td>ZAK</td>\n",
       "      <td>Q9NYL2</td>\n",
       "      <td>ZAK</td>\n",
       "      <td>Q9NYL2</td>\n",
       "      <td>MAP3K20</td>\n",
       "      <td>T161</td>\n",
       "      <td>ASRFHNHtTHMSLVG</td>\n",
       "      <td>2q31.1</td>\n",
       "      <td>M3K20_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10979</td>\n",
       "      <td>MEKK4</td>\n",
       "      <td>Q9Y6R4</td>\n",
       "      <td>MEKK4</td>\n",
       "      <td>Q9Y6R4</td>\n",
       "      <td>MAP3K4</td>\n",
       "      <td>T1494</td>\n",
       "      <td>KLKNNAQtMPGEVNS</td>\n",
       "      <td>6q26</td>\n",
       "      <td>M3K4_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10980</td>\n",
       "      <td>NuaK1</td>\n",
       "      <td>O60285</td>\n",
       "      <td>CASP6</td>\n",
       "      <td>P55212</td>\n",
       "      <td>CASP6</td>\n",
       "      <td>S257</td>\n",
       "      <td>TLVNRKVsQRRVDFC</td>\n",
       "      <td>4q25</td>\n",
       "      <td>CASP6_HUMAN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10981 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      KINASE_GENE_NAME KIN_UNIPROT_ID SUBSTRATE_NAME SUB_UNIPROT_ID  \\\n",
       "0                  HRI         Q9BQI3     eIF2-alpha         P05198   \n",
       "1                  HRI         Q9BQI3     eIF2-alpha         P05198   \n",
       "2                  PKR         P19525     eIF2-alpha         P05198   \n",
       "3                  PKR         P19525     eIF2-alpha         P05198   \n",
       "4                 P38A         Q16539     eIF2-alpha         P05198   \n",
       "...                ...            ...            ...            ...   \n",
       "10976              ZAK         Q9NYL2            ZAK         Q9NYL2   \n",
       "10977              ZAK         Q9NYL2            ZAK         Q9NYL2   \n",
       "10978              ZAK         Q9NYL2            ZAK         Q9NYL2   \n",
       "10979            MEKK4         Q9Y6R4          MEKK4         Q9Y6R4   \n",
       "10980            NuaK1         O60285          CASP6         P55212   \n",
       "\n",
       "      SUB_GENE_NAME SUB_MOD_RSD        SITE_7_AA  CHR_LOC SUB_ENTRY_NAME  \n",
       "0            EIF2S1         S52  MILLSELsRRRIRSI  14q23.3     IF2A_HUMAN  \n",
       "1            EIF2S1         S49  IEGMILLsELSRRRI  14q23.3     IF2A_HUMAN  \n",
       "2            EIF2S1         S49  IEGMILLsELSRRRI  14q23.3     IF2A_HUMAN  \n",
       "3            EIF2S1         S52  MILLSELsRRRIRSI  14q23.3     IF2A_HUMAN  \n",
       "4            EIF2S1         S52  MILLSELsRRRIRSI  14q23.3     IF2A_HUMAN  \n",
       "...             ...         ...              ...      ...            ...  \n",
       "10976       MAP3K20        T162  SRFHNHTtHMSLVGT   2q31.1    M3K20_HUMAN  \n",
       "10977       MAP3K20        S165  HNHTTHMsLVGTFPW   2q31.1    M3K20_HUMAN  \n",
       "10978       MAP3K20        T161  ASRFHNHtTHMSLVG   2q31.1    M3K20_HUMAN  \n",
       "10979        MAP3K4       T1494  KLKNNAQtMPGEVNS     6q26     M3K4_HUMAN  \n",
       "10980         CASP6        S257  TLVNRKVsQRRVDFC     4q25    CASP6_HUMAN  \n",
       "\n",
       "[10981 rows x 9 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#...and merge it with our original dataframe so that UniProt Entries are included.\n",
    "df = pd.merge(df,df2[['SUB_UNIPROT_ID','SUB_ENTRY_NAME']],on='SUB_UNIPROT_ID', how='inner')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PHOSPHO_ID</th>\n",
       "      <th>KINASE_GENE_NAME</th>\n",
       "      <th>KIN_UNIPROT_ID</th>\n",
       "      <th>SUBSTRATE_NAME</th>\n",
       "      <th>SUB_UNIPROT_ID</th>\n",
       "      <th>SUB_GENE_NAME</th>\n",
       "      <th>SUB_MOD_RSD</th>\n",
       "      <th>SITE_7_AA</th>\n",
       "      <th>CHR_LOC</th>\n",
       "      <th>SUB_ENTRY_NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>HRI</td>\n",
       "      <td>Q9BQI3</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>S52</td>\n",
       "      <td>MILLSELsRRRIRSI</td>\n",
       "      <td>14q23.3</td>\n",
       "      <td>IF2A_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>HRI</td>\n",
       "      <td>Q9BQI3</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>S49</td>\n",
       "      <td>IEGMILLsELSRRRI</td>\n",
       "      <td>14q23.3</td>\n",
       "      <td>IF2A_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>PKR</td>\n",
       "      <td>P19525</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>S49</td>\n",
       "      <td>IEGMILLsELSRRRI</td>\n",
       "      <td>14q23.3</td>\n",
       "      <td>IF2A_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>PKR</td>\n",
       "      <td>P19525</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>S52</td>\n",
       "      <td>MILLSELsRRRIRSI</td>\n",
       "      <td>14q23.3</td>\n",
       "      <td>IF2A_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>P38A</td>\n",
       "      <td>Q16539</td>\n",
       "      <td>eIF2-alpha</td>\n",
       "      <td>P05198</td>\n",
       "      <td>EIF2S1</td>\n",
       "      <td>S52</td>\n",
       "      <td>MILLSELsRRRIRSI</td>\n",
       "      <td>14q23.3</td>\n",
       "      <td>IF2A_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10976</td>\n",
       "      <td>10977</td>\n",
       "      <td>ZAK</td>\n",
       "      <td>Q9NYL2</td>\n",
       "      <td>ZAK</td>\n",
       "      <td>Q9NYL2</td>\n",
       "      <td>MAP3K20</td>\n",
       "      <td>T162</td>\n",
       "      <td>SRFHNHTtHMSLVGT</td>\n",
       "      <td>2q31.1</td>\n",
       "      <td>M3K20_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10977</td>\n",
       "      <td>10978</td>\n",
       "      <td>ZAK</td>\n",
       "      <td>Q9NYL2</td>\n",
       "      <td>ZAK</td>\n",
       "      <td>Q9NYL2</td>\n",
       "      <td>MAP3K20</td>\n",
       "      <td>S165</td>\n",
       "      <td>HNHTTHMsLVGTFPW</td>\n",
       "      <td>2q31.1</td>\n",
       "      <td>M3K20_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10978</td>\n",
       "      <td>10979</td>\n",
       "      <td>ZAK</td>\n",
       "      <td>Q9NYL2</td>\n",
       "      <td>ZAK</td>\n",
       "      <td>Q9NYL2</td>\n",
       "      <td>MAP3K20</td>\n",
       "      <td>T161</td>\n",
       "      <td>ASRFHNHtTHMSLVG</td>\n",
       "      <td>2q31.1</td>\n",
       "      <td>M3K20_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10979</td>\n",
       "      <td>10980</td>\n",
       "      <td>MEKK4</td>\n",
       "      <td>Q9Y6R4</td>\n",
       "      <td>MEKK4</td>\n",
       "      <td>Q9Y6R4</td>\n",
       "      <td>MAP3K4</td>\n",
       "      <td>T1494</td>\n",
       "      <td>KLKNNAQtMPGEVNS</td>\n",
       "      <td>6q26</td>\n",
       "      <td>M3K4_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10980</td>\n",
       "      <td>10981</td>\n",
       "      <td>NuaK1</td>\n",
       "      <td>O60285</td>\n",
       "      <td>CASP6</td>\n",
       "      <td>P55212</td>\n",
       "      <td>CASP6</td>\n",
       "      <td>S257</td>\n",
       "      <td>TLVNRKVsQRRVDFC</td>\n",
       "      <td>4q25</td>\n",
       "      <td>CASP6_HUMAN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10981 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       PHOSPHO_ID KINASE_GENE_NAME KIN_UNIPROT_ID SUBSTRATE_NAME  \\\n",
       "0               1              HRI         Q9BQI3     eIF2-alpha   \n",
       "1               2              HRI         Q9BQI3     eIF2-alpha   \n",
       "2               3              PKR         P19525     eIF2-alpha   \n",
       "3               4              PKR         P19525     eIF2-alpha   \n",
       "4               5             P38A         Q16539     eIF2-alpha   \n",
       "...           ...              ...            ...            ...   \n",
       "10976       10977              ZAK         Q9NYL2            ZAK   \n",
       "10977       10978              ZAK         Q9NYL2            ZAK   \n",
       "10978       10979              ZAK         Q9NYL2            ZAK   \n",
       "10979       10980            MEKK4         Q9Y6R4          MEKK4   \n",
       "10980       10981            NuaK1         O60285          CASP6   \n",
       "\n",
       "      SUB_UNIPROT_ID SUB_GENE_NAME SUB_MOD_RSD        SITE_7_AA  CHR_LOC  \\\n",
       "0             P05198        EIF2S1         S52  MILLSELsRRRIRSI  14q23.3   \n",
       "1             P05198        EIF2S1         S49  IEGMILLsELSRRRI  14q23.3   \n",
       "2             P05198        EIF2S1         S49  IEGMILLsELSRRRI  14q23.3   \n",
       "3             P05198        EIF2S1         S52  MILLSELsRRRIRSI  14q23.3   \n",
       "4             P05198        EIF2S1         S52  MILLSELsRRRIRSI  14q23.3   \n",
       "...              ...           ...         ...              ...      ...   \n",
       "10976         Q9NYL2       MAP3K20        T162  SRFHNHTtHMSLVGT   2q31.1   \n",
       "10977         Q9NYL2       MAP3K20        S165  HNHTTHMsLVGTFPW   2q31.1   \n",
       "10978         Q9NYL2       MAP3K20        T161  ASRFHNHtTHMSLVG   2q31.1   \n",
       "10979         Q9Y6R4        MAP3K4       T1494  KLKNNAQtMPGEVNS     6q26   \n",
       "10980         P55212         CASP6        S257  TLVNRKVsQRRVDFC     4q25   \n",
       "\n",
       "      SUB_ENTRY_NAME  \n",
       "0         IF2A_HUMAN  \n",
       "1         IF2A_HUMAN  \n",
       "2         IF2A_HUMAN  \n",
       "3         IF2A_HUMAN  \n",
       "4         IF2A_HUMAN  \n",
       "...              ...  \n",
       "10976    M3K20_HUMAN  \n",
       "10977    M3K20_HUMAN  \n",
       "10978    M3K20_HUMAN  \n",
       "10979     M3K4_HUMAN  \n",
       "10980    CASP6_HUMAN  \n",
       "\n",
       "[10981 rows x 10 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Finally, we insert a a PHOSPHO_ID column into our dataframe \n",
    "df.insert(0, 'PHOSPHO_ID', range(1, 1 + len(df)))\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Writing our final dataframe into a CSV file\n",
    "df.to_csv('Phosphosite_Table.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lastly, we check through our final file for any errors that either were missed or not fixable through code and correct them manually. An examples of this include correcting the Substrate Gene Names such as 'SEPT1' and 'Oct1' back to their actual names as Excel read these genes and others with similar names as dates instead of gene names. Once things have been thoroughly checked our Phosphosite file is ready to be implemented to our database.    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
